File: src/Java/BubbleSort_end.java
import java.util.Arrays;

public class BubbleSort_end { // Class name matches the filename (if you keep it as BubbleSort_end.java)

    void bubbleSort(int arr[]) {
        int n = arr.length;
        boolean swapped;
        for (int i = 0; i < n - 1; i++) {
            swapped = false;
            for (int j = 0; j < n - i - 1; j++) {
                if (arr[j] > arr[j + 1]) {
                    // swap arr[j+1] and arr[j]
                    int temp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = temp;
                    swapped = true;
                }
            }
            // IF no two elements were swapped by inner loop, then break
            if (!swapped)
                break;
        }
    }

    // Method to print the array
    void printArray(int arr[]) {
        System.out.println(Arrays.toString(arr));
    }

    // Driver method to test above
    public static void main(String args[]) {
        // THIS IS THE LINE THAT NEEDS TO CHANGE:
        BubbleSort_end ob = new BubbleSort_end(); // Create an object of the CURRENT class
        
        int arr1[] = {64, 34, 25, 12, 22, 11, 90};
        System.out.println("Java Example 1:");
        System.out.print("Original array: ");
        ob.printArray(arr1);
        ob.bubbleSort(arr1);
        System.out.print("Sorted array: ");
        ob.printArray(arr1);
        System.out.println("--------------------");

        int arr2[] = {5, 1, 4, 2, 8, 3, 7, 6};
        System.out.println("Java Example 2:");
        System.out.print("Original array: ");
        ob.printArray(arr2);
        ob.bubbleSort(arr2);
        System.out.print("Sorted array: ");
        ob.printArray(arr2);
        System.out.println("--------------------");

        int emptyArr[] = {};
        System.out.println("Java Example (Empty Array):");
        System.out.print("Original array: ");
        ob.printArray(emptyArr);
        ob.bubbleSort(emptyArr);
        System.out.print("Sorted array: ");
        ob.printArray(emptyArr);
        System.out.println("--------------------");

        int singleElementArr[] = {42};
        System.out.println("Java Example (Single Element Array):");
        System.out.print("Original array: ");
        ob.printArray(singleElementArr);
        ob.bubbleSort(singleElementArr);
        System.out.print("Sorted array: ");
        ob.printArray(singleElementArr);
    }
}

File: src/C/bubble_sort_end.c
#include <stdio.h>
#include <stdbool.h>

// Function to swap two elements
void swap(int *xp, int *yp) {
    int temp = *xp;
    *xp = *yp;
    *yp = temp;
}

// Function to implement bubble sort
void bubbleSort(int arr[], int n) {
    if (n <= 0) return; // Explicitly handle empty or invalid size
    int i, j;
    bool swapped;
    for (i = 0; i < n - 1; i++) {
        swapped = false;
        for (j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(&arr[j], &arr[j + 1]);
                swapped = true;
            }
        }
        if (swapped == false)
            break;
    }
}

// Function to print an array
void printArray(int arr[], int size) {
    if (size <= 0) { // Handle empty or invalid size
        printf("\n");
        return;
    }
    int i;
    for (i = 0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

// Driver program to test above functions
int main() {
    int arr1[] = {64, 34, 25, 12, 22, 11, 90};
    int n1 = sizeof(arr1) / sizeof(arr1[0]);
    printf("C Example 1:\n");
    printf("Original array: ");
    printArray(arr1, n1);
    bubbleSort(arr1, n1);
    printf("Sorted array: ");
    printArray(arr1, n1);
    printf("--------------------\n");

    int arr2[] = {5, 1, 4, 2, 8, 3, 7, 6};
    int n2 = sizeof(arr2) / sizeof(arr2[0]);
    printf("C Example 2:\n");
    printf("Original array: ");
    printArray(arr2, n2);
    bubbleSort(arr2, n2);
    printf("Sorted array: ");
    printArray(arr2, n2);
    printf("--------------------\n");
    
    // For the empty case, pass NULL or a valid pointer with size 0
    // Passing NULL is conceptually clearer for "no array"
    // int *empty_arr = NULL; // Option A: NULL pointer
    int dummy_arr[1]; // Option B: A valid pointer, but size 0 passed
    int n_empty = 0; 
    printf("C Example (Empty Array):\n");
    printf("Original array: ");
    // printArray(empty_arr, n_empty); // if using Option A (NULL)
    printArray(dummy_arr, n_empty); // if using Option B
    // bubbleSort(empty_arr, n_empty); // if using Option A (NULL)
    bubbleSort(dummy_arr, n_empty); // if using Option B
    printf("Sorted array: ");
    // printArray(empty_arr, n_empty); // if using Option A (NULL)
    printArray(dummy_arr, n_empty); // if using Option B
    printf("--------------------\n");

    int single_arr[] = {42};
    int n_single = sizeof(single_arr)/sizeof(single_arr[0]);
    printf("C Example (Single Element Array):\n");
    printf("Original array: ");
    printArray(single_arr, n_single);
    bubbleSort(single_arr, n_single);
    printf("Sorted array: ");
    printArray(single_arr, n_single);

    return 0;
}